### DISCUSSION TITLE
How would you find all git repository in a directory ?

### DISCUSSION THREAD

**Discussion by PierreAntoineGuillaume** [sharkdp/fd#1323] [Q&A] [answered]
I have a `find` command that helps me iterate over all git repository : 

```bash
find $WORKSPACE -type d -exec test -e '{}/.git' ';' -print -prune
```

how would I replicate that with `fd` ?

On a side note, what I did is better that the `find` command in terms of performance : 

```
fd '.git$' "$WORKSPACE" --hidden
```

But I don't know if I can improve the performance by telling `fd` not to recurse in folders where it finds .git subdirs

**Comment by tavianator** [comment, marked as answer]
You can use `--prune` so `fd` doesn't recurse inside the `.git` directories themselves:

```console
$ fd -u '^\.git$' --prune -X printf '%s\n' '{//}'
```

That doesn't prune everything but I suspect it's hard to beat in practice.  You might also find that this GNU find command line is much faster than using `-exec test`:

```console
$ find -name .git -printf '%h\n' -prune
$ # Or, if your find doesn't have -printf
$ find -name .git -prune | sed 's|/.git$||'
```

**Reply by Machine-Maker** [reply]
Is there a way to not recurse inside the parent of the `.git` directory? To exclude repos within repos?

**Reply by PierreAntoineGuillaume** [reply]
> Is there a way to not recurse inside the parent of the `.git` directory? To exclude repos within repos?

I'm not sure what you mean.

Could you rephrase ?

**Reply by Machine-Maker** [reply]
So it matches against `.git` directories, and then `--prune` prevents anything inside those `.git` directories, but I also want to exclude everything inside the parent directory of `.git`. 

So you have
```
Projects/
	Thing/
		.git/
		sub-repo/
			.git/
```
I don't want to match against that inner `.git` because a parent of that `.git` is already matched

**Reply by PierreAntoineGuillaume** [reply]
So the original `find` command worked !

Sorry about the delay, can't find anything to help you !

**Reply by PierreAntoineGuillaume** [reply]
@Machine-Maker if you can afford to compile some rust code yourself, this lib depends on `ignore`, a subcrate of ripgrep, which in turn depends on `BurntSushi/Walkdir`.


This is a sample of some poorly written code : 

```rust
let mut iter = WalkDir::new(std::env::var("PWD").unwrap()).into_iter();
loop {
    match iter.next() {
        None => break,
        Some(Err(_)) => break,
        Some(Ok(entry)) => {
            if entry.path().is_dir() {
                if entry.path().starts_with(".") {
                    iter.skip_current_dir();
                    continue;
                }
                let pathdir = entry.path().display().to_string();
                let git_path = format!("{pathdir}/.git");
                if Path::new(&git_path).exists() {
                    println!("{pathdir}");
                    iter.skip_current_dir();
                }
            }
        }
    }
}
```

In this example, the code would report all directories containing a .git repository, without recursing within it.

**Reply by toshmukhamedov** [reply]
<img width="763" alt="image" src="https://github.com/sharkdp/fd/assets/79999469/6b1a5a69-f639-408d-b1c2-c93d137441c4">

Is there any better way to do this?

**Reply by PierreAntoineGuillaume** [reply]
> <img width="763" alt="image" src="https://github.com/sharkdp/fd/assets/79999469/6b1a5a69-f639-408d-b1c2-c93d137441c4">
> 
> Is there any better way to do this?

https://github.com/sharkdp/fd/discussions/1323#discussioncomment-8423942

There may be a small issue with the algorithm, i can check at a later time that entry.path().starts_with is the correct thing to do to avoid descend into hidden dirs

**Reply by PierreAntoineGuillaume** [reply]
@toshmukhamedov that does it for you ?
